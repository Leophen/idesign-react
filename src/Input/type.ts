export interface InputProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 占位符
   * @default 请输入
   */
  placeholder?: string;
  /**
   * 输入框固定值（受控）
   */
  value?: string | number;
  /**
   * 输入框默认值（非受控）
   */
  defaultValue?: string | number;
  /**
   * 是否禁用输入框
   * @default false
   */
  disabled?: boolean;
  /**
   * 输入框是否只读
   * @default false
   */
  readonly?: boolean;
  /**
   * 是否隐藏数字输入框后缀按钮
   * @default false
   */
  hideNumberBtn?: boolean;
  /**
   * 是否聚焦时全选
   * @default false
   */
  selectAll?: boolean;
  /**
   * 输入框尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 输入框状态
   */
  status?: 'success' | 'warning' | 'error';
  /**
   * 输入框底部提示
   */
  tips?: string;
  /**
   * 用户最多可以输入的字符个数
   */
  maxLength?: number;
  /**
   * 是否可清空
   * @default false
   */
  clearable?: boolean;
  /**
   * 输入框类型
   * @default text
   */
  type?: 'text' | 'password' | 'number';
  /**
   * 数字输入框滑块数值变化速率
   * @default default
   */
  speed?: 'slow' | 'default' | 'fast';
  /**
   * 数字输入框最大值
   */
  maxNumber?: number;
  /**
   * 数字输入框最小值
   */
  minNumber?: number;
  /**
   * 数字输入框清空值时是否转为最小值
   * @default false
   */
  minNumberLock?: boolean;
  /**
   * 数字输入框保留几位小数
   * @default 0
   */
  precision?: number;
  /**
   * 数字输入框数值变化间隔
   * @default 1
   */
  step?: number;
  /**
   * 组件前置图标名
   */
  prefixIcon?: string;
  /**
   * 组件后置图标名
   */
  suffixIcon?: string;
  /**
   * 组件前置图标类名
   */
  prefixIconClass?: string;
  /**
   * 组件后置图标类名
   */
  suffixIconClass?: string;
  /**
   * 输入框值发生变化时触发
   */
  onChange?: (value: string, context?: { e?: React.ChangeEvent<HTMLInputElement> }) => void;
  /**
   * 输入框聚焦时触发
   */
  onFocus?: (value: string, context?: { e?: React.FocusEvent<HTMLInputElement> }) => void;
  /**
   * 输入框失焦时触发
   */
  onBlur?: (value: string, context?: { e?: React.FocusEvent<HTMLInputElement> }) => void;
  /**
   * 键盘按下时触发
   */
  onKeyDown?: (value: string, context: { e: React.KeyboardEvent<HTMLInputElement> }) => void;
  /**
   * 键盘按下回车键时触发
   */
  onEnter?: (value: string, context: { e: React.KeyboardEvent<HTMLInputElement> }) => void;
  /**
   * 释放键盘时触发
   */
  onKeyUp?: (value: string, context: { e: React.KeyboardEvent<HTMLDivElement> }) => void;
  /**
   * 清空按钮点击时触发
   */
  onClear?: (context: { e: React.MouseEvent<HTMLDivElement> }) => void;
  /**
   * 数字输入框移动滑块时触发
   */
  onMove?: (value: string, context?: { e?: React.ChangeEvent<HTMLInputElement> }) => void;
  /**
   * 数字输入框滑块失焦时触发
   */
  onMoveUp?: (value: string) => void;
  /**
   * 点击前置图标触发事件
   */
  clickPrefixIcon?: (context: { e: React.MouseEvent<HTMLDivElement> }) => void;
  /**
   * 点击后置图标触发事件
   */
  clickSuffixIcon?: (context: { e: React.MouseEvent<HTMLDivElement> }) => void;
}

export interface InputGroupProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 输入框组合前缀
   */
  prefixContent?: React.ReactNode;
  /**
   * 输入框组合后缀
   */
  suffixContent?: React.ReactNode;
  /**
   * 点击前缀触发事件
   */
  clickPrefix?: (context: { e: React.MouseEvent<HTMLDivElement> }) => void;
  /**
   * 点击后缀触发事件
   */
  clickSuffix?: (context: { e: React.MouseEvent<HTMLDivElement> }) => void;
}
