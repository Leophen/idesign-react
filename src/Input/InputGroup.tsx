import React, { useEffect, useState, useRef } from 'react';
import classNames from 'classnames';
import './index.scss';
import { InputGroupProps } from './type'

const InputGroup: React.FC<InputGroupProps> = (props) => {
  const {
    className,
    children,
    style,
    prefixContent,
    suffixContent,
    clickPrefix,
    clickSuffix
  } = props;

  const [contentHeight, setContentHeight] = useState(0);
  const groupNode = useRef(null);

  useEffect(() => {
    let maxHeight = 0;
    const groupChild = groupNode.current && (groupNode.current as HTMLElement).childNodes;
    groupChild &&
      Array.from(groupChild).map((item: any) => {
        if (
          item.className !== 'i-input__group-prefix' &&
          item.className !== 'i-input__group-suffix' &&
          item.offsetHeight > maxHeight
        ) {
          maxHeight = item.offsetHeight;
        }
      });
    setContentHeight(maxHeight);
  });

  const handleClickWith = (location: 'pre' | 'suf', e: React.MouseEvent<HTMLDivElement>) => {
    e.persist();
    e.stopPropagation();
    if (clickPrefix && location === 'pre') {
      clickPrefix(e as any);
      return;
    }
    if (clickSuffix && location === 'suf') {
      clickSuffix(e as any);
      return;
    }
  };

  return (
    <div
      className={classNames('i-input__group', className)}
      style={{ ...(style || {}), ...{ height: contentHeight } }}
      ref={groupNode}
    >
      {prefixContent && (
        <div
          className={classNames('i-input__group-prefix', clickPrefix && 'i-input__group-cursor')}
          onClick={(e: React.MouseEvent<HTMLDivElement>) => handleClickWith('pre', e)}
        >
          {prefixContent}
        </div>
      )}
      {children}
      {suffixContent && (
        <div
          className={classNames('i-input__group-suffix', clickSuffix && 'i-input__group-cursor')}
          onClick={(e: React.MouseEvent<HTMLDivElement>) => handleClickWith('suf', e)}
        >
          {suffixContent}
        </div>
      )}
    </div>
  );
};

export default InputGroup;
