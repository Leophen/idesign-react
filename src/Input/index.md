---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 4
---

# Input 输入框

用于字符输入的单行输入框。

## 基本用法

基础的单行输入框，可通过 `defaultValue` 设置默认值，通过 `value` 属性设置受控值，通过 `onChange` 设置值改变时触发的操作：

```tsx
import React, { useState } from 'react'
import { Input } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('这是默认值')

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <Input />
      <h4>有默认值</h4>
      <Input defaultValue={value} />
      <h4>有受控值</h4>
      <Input value={value} />
      <h4>一般用法</h4>
      <Input value={value} onChange={handleChange} />
    </>
  )
}

export default App
```

## 禁用状态

通过 `disabled` 属性指定是否禁用输入框。

```tsx
import React from 'react';
import { Input } from 'idesign-react';

const App = () => {
  return <Input disabled />;
};

export default App;
```

## 只读状态

通过 `readonly` 属性指定是否为只读状态。

```tsx
import React from 'react';
import { Input } from 'idesign-react';

const App = () => {
  return <Input readonly />;
};

export default App;
```

## 聚焦全选

通过 `selectAll` 属性指定聚焦状态时自动全选。

```tsx
import React from 'react';
import { Input } from 'idesign-react';

const App = () => {
  return <Input selectAll />;
};

export default App;
```

## 不同尺寸

通过 `size` 属性指定不同尺寸的输入框，默认为 `medium`。

```tsx
import React from 'react';
import { Input } from 'idesign-react';

const App = () => {
  return (
    <>
      <Input size="small" />
      <br />
      <Input size="medium" />
      <br />
      <Input size="large" />
    </>
  );
};

export default App;
```

## 不同状态

通过 `status` 属性指定不同状态的输入框，通过 `tips` 属性指定 底部提示信息。

```tsx
import React from 'react';
import { Input } from 'idesign-react';

const App = () => {
  return (
    <>
      <Input placeholder="成功状态" status="success" />
      <br />
      <Input placeholder="警告状态" status="warning" />
      <br />
      <Input placeholder="错误状态" status="error" />
      <br />
      <Input placeholder="成功状态" status="success" tips="成功状态提示" />
      <Input placeholder="警告状态" status="warning" tips="警告状态提示" />
      <Input placeholder="错误状态" status="error" tips="错误状态提示" />
    </>
  );
};

export default App;
```

## 限制字数

通过 `maxLength` 属性指定最多可以输入的字符个数。

```tsx
import React, { useState } from 'react';
import { Input } from 'idesign-react';

const App = () => {
  const [value, setValue] = useState('iDesign');

  return (
    <>
      <Input maxLength={15} />
      <br />
      <Input
        value={value}
        maxLength={20}
        onChange={(value) => {
          console.log(value);
          setValue(value);
        }}
      />
    </>
  );
};

export default App;
```

## 密码输入框

通过将 `type` 属性指定为 `password` 切换为密码输入框。

```tsx
import React from 'react';
import { Input } from 'idesign-react';

const App = () => {
  return <Input type="password" />;
};

export default App;
```

## 高级数字输入框

通过将 `type` 属性指定为 `number` 切换为数字滑块输入框：右侧显示数控按钮、悬浮左边缘出现鼠标数控滑块、可通过键盘上下键微调数值。

另外，可通过 `speed` 属性值设置滑块移动数值变化速率，通过 `step` 设置数值变化间隔，通过 `precision` 设置数值保留几位有效数字，通过 `hideNumberBtn` 设置隐藏数字输入框后缀按钮。

```tsx
import React, { useState } from 'react';
import { Input } from 'idesign-react';

const App = () => {
  const [value, setValue] = useState('1');

  const handleChange = (val) => {
    setValue(val)
  }

  return (
    <>
      <h4>默认数字输入框</h4>
      <Input type="number" />
      <h4>隐藏调整控件的数字输入框</h4>
      <Input type="number" hideNumberBtn />
      <h4>受控且有最大最小值的数字输入框</h4>
      <Input value={value} type="number" minNumber={1} maxNumber={50} onChange={handleChange} />
      <h4>受控且有最大最小值并锁定空值的数字输入框</h4>
      <Input value={value} type="number" minNumber={1} maxNumber={50} minNumberLock onChange={handleChange} />
      <h4>慢速拖拽控件且有最大值的数字输入框</h4>
      <Input
        placeholder="speed 设为 slow + 最大值 100"
        speed="slow"
        type="number"
        maxNumber={100}
      />
      <h4>快速拖拽控件且有最小值的数字输入框</h4>
      <Input placeholder="speed 设为 fast + 最小值 0" speed="fast" type="number" minNumber={0} />
      <h4>设置精度及数字间隔的数字输入框</h4>
      <Input
        placeholder="保留 2 位有效数字 + 间隔 0.5"
        type="number"
        precision={2}
        step={0.5}
        onChange={(value) => {
          console.log(value);
        }}
      />
      <h4>带前后缀图标的数字输入框</h4>
      <Input
        type="number"
        prefixIcon="TagCollection"
        suffixIcon="Search"
        clickSuffixIcon={() => {
          alert('Search');
        }}
      />
    </>
  );
};

export default App;
```

## 带清空按钮的输入框

通过 `clearable` 属性或 `onClear` 显示清空按钮，实现一键清空输入框内容。

```tsx
import React, { useState } from 'react';
import { Input } from 'idesign-react';

const App = () => {
  const [value1, setValue1] = useState('iDesign1');
  const [value2, setValue2] = useState('iDesign2');

  return (
    <>
      <Input
        value={value1}
        clearable
        onChange={(val) => {
          setValue1(val);
        }}
      />
      <br />
      <Input
        value={value2}
        onChange={(val) => {
          setValue2(val);
        }}
        onClear={(e) => {
          console.log('onClear', e);
        }}
      />
    </>
  );
};

export default App;
```

## 带内置图标的输入框

通过 `prefixIcon` 或 `suffixIcon` 来指定输入框前后内置图标。

```tsx
import React, { useState } from 'react';
import { Input } from 'idesign-react';

const App = () => {
  const [value, setValue] = useState('iDesign');

  return (
    <>
      <Input prefixIcon="Search" />
      <br />
      <Input suffixIcon="Calendar" />
      <br />
      <Input
        value={value}
        clearable
        suffixIcon="Search"
        onChange={(val) => {
          setValue(val);
        }}
      />
      <br />
      <Input
        prefixIcon="Search"
        suffixIcon="Calendar"
        clickPrefixIcon={(e) => console.log(e)}
        clickSuffixIcon={(e) => console.log(e)}
      />
    </>
  );
};

export default App;
```

## 组合输入框

通过 `Input.Group` 包裹不同内容组成单行的组合输入框，可结合 [Grid](./grid) 组件来控制宽度占比。

```tsx
import React from 'react';
import { Input } from 'idesign-react';
import { Grid } from 'idesign-react';
import { Button } from 'idesign-react';
import { Icon } from 'idesign-react';

const App = () => {
  const prefixIcon = <Icon name="ShoppingCart" />;
  const suffixIcon = <Icon name="Search" />;

  return (
    <>
      <h3>输入框 + 输入框</h3>
      <Input.Group>
        <Input />
        <Input />
      </Input.Group>
      <br />
      <Input.Group>
        <Grid>
          <Grid.Item width={120}>
            <Input />
          </Grid.Item>
          <Grid.Item>
            <Input />
          </Grid.Item>
        </Grid>
      </Input.Group>
      <br />
      <Input.Group>
        <Grid>
          <Grid.Item span={4}>
            <Input />
          </Grid.Item>
          <Grid.Item span={8}>
            <Input />
          </Grid.Item>
          <Grid.Item span={12}>
            <Input />
          </Grid.Item>
        </Grid>
      </Input.Group>

      <h3>输入框 + 按钮</h3>
      <Input.Group>
        <Input />
        <Button style={{ marginLeft: '10px' }}>搜索</Button>
      </Input.Group>

      <h3>输入框 + 前后缀</h3>
      <Input.Group prefixContent="http://">
        <Input />
      </Input.Group>
      <br />
      <Input.Group prefixContent="http://" suffixContent=".com">
        <Input />
      </Input.Group>
      <br />
      <Input.Group
        prefixContent={prefixIcon}
        suffixContent={suffixIcon}
        clickPrefix={(e) => console.log(e)}
        clickSuffix={(e) => console.log(e)}
      >
        <Input />
      </Input.Group>
    </>
  );
};

export default App;
```

## 触发事件

一系列的事件触发操作。

```tsx
import React, { useState } from 'react';
import { Input } from 'idesign-react';

const App = () => {
  const [value, setValue] = useState('iDesign');

  return (
    <>
      <h3>onChange 输入时触发</h3>
      <Input
        onChange={(val, e) => {
          console.log('onChange', val, e);
        }}
      />
      <h3>onFocus 聚焦时触发</h3>
      <Input
        onFocus={(val, e) => {
          console.log('onFocus', val, e);
        }}
      />
      <h3>onBlur 失焦时触发</h3>
      <Input
        onBlur={(val, e) => {
          console.log('onBlur', val, e);
        }}
      />
      <h3>onKeyDown 键盘按下时触发</h3>
      <Input
        onKeyDown={(val, e) => {
          console.log('onKeyDown', val, e);
        }}
      />
      <h3>onEnter 键盘按下回车时触发</h3>
      <Input
        onEnter={(val, e) => {
          console.log('onEnter', val, e);
        }}
      />
      <h3>onKeyUp 键盘释放时触发</h3>
      <Input
        onKeyUp={(val, e) => {
          console.log('onKeyUp', val, e);
        }}
      />
      <h3>onClear 清空按钮点击时触发</h3>
      <Input
        value={value}
        onChange={(val) => {
          setValue(val);
        }}
        onClear={(e) => {
          console.log('onClear', e);
        }}
      />
    </>
  );
};

export default App;
```

## Input API

<API hideTitle />

## InputGroup API

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| children | 按钮内容 | `ReactNode` | `--` |
| className | 类名 | `string` | `--` |
| style | 自定义样式 | `CSSProperties` | `--` |
| prefixContent | 输入框组合前缀 | `ReactNode` | `--` |
| clickPrefix | 点击前缀触发事件 | `(context: { e: React.MouseEvent<HTMLDivElement> }) => void` | `--` |
| clickSuffix | 点击后缀触发事件 | `(context: { e: React.MouseEvent<HTMLDivElement> }) => void` | `--` |
