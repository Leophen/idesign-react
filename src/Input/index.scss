@mixin settingStatus($color) {
  &,
  &:hover,
  &:focus-within {
    border-color: $color;
  }
}

.i-input {
  position: relative;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  width: 100%;
  height: 32px;
  padding: 0 8px;
  overflow: hidden;
  color: var(--i-font-2);
  font-size: 14px;
  line-height: 32px;
  border: 1px solid var(--i-border);
  border-radius: 4px;
  outline: none;
  visibility: visible;
  cursor: text;
  transition: all 0.2s cubic-bezier(0.38, 0, 0.24, 1);
  .i-icon {
    margin-left: 6px;
    color: var(--i-font-4) !important;
    font-size: inherit !important;
    cursor: pointer;
    transition: all 0.1s ease-out;
    &:hover {
      color: var(--i-font-3) !important;
    }
  }
  .icon-TipCloseFill {
    // display: none;
    opacity: 0;
  }
  .i-input-prefix-icon,
  .i-input-suffix-icon {
    cursor: text;
    &:hover {
      color: var(--i-font-4) !important;
    }
    &.i-input-icon-cursor {
      cursor: pointer;
      &:hover {
        color: var(--i-font-3) !important;
      }
    }
  }
  .i-input-prefix-icon {
    margin-right: 6px;
    margin-left: 0;
  }
  &.i-input--size-small {
    height: 24px;
    padding: 0 6px;
    font-size: 12px;
    line-height: 24px;
    .i-input__inner {
      height: 24px;
    }
    .i-icon {
      margin-left: 5px;
    }
    .i-input-prefix-icon {
      margin-right: 5px;
      margin-left: 0;
    }
    .i-input-number-button {
      width: 12px;
      .i-icon {
        margin-left: 6px;
        padding-right: 3px;
        padding-left: 3px;
      }
    }
  }
  &.i-input--size-large {
    height: 40px;
    padding: 0 10px;
    font-size: 16px;
    line-height: 40px;
    .i-input__inner {
      height: 40px;
    }
    .i-icon {
      margin-left: 7px;
    }
    .i-input-prefix-icon {
      margin-right: 7px;
      margin-left: 0;
    }
    .i-input-number-button {
      width: 16px;
      .i-icon {
        margin-left: 10px;
        padding-right: 5px;
        padding-left: 5px;
      }
    }
  }
  &:hover {
    border-color: var(--i-border-hover);
    .icon-TipCloseFill {
      // display: block;
      opacity: 1;
    }
  }
  &:focus-within {
    border-color: var(--i-primary);
  }
  &.i-input-is-disabled {
    background-color: var(--i-bg-disabled);
    .i-input__inner {
      color: var(--i-font-disabled);
    }
    &,
    * {
      cursor: not-allowed;
    }
    &:hover {
      border-color: var(--i-border);
    }
  }
  &.i-input-is-readonly {
    &,
    * {
      cursor: default;
    }
  }
  &.i-input--status-success {
    @include settingStatus(var(--i-success));
  }
  &.i-input--status-warning {
    @include settingStatus(var(--i-warning));
  }
  &.i-input--status-error {
    @include settingStatus(var(--i-error));
  }
}

.i-input--type-password {
  .i-input__inner {
    letter-spacing: 2px;
    &::placeholder {
      letter-spacing: 0;
    }
  }
}

.i-input--type-number {
  padding-right: 0;
  user-select: none;
  .i-input__inner {
    padding-right: 8px;
    &:focus {
      user-select: text;
    }
  }
  .i-input-suffix-icon {
    position: relative;
    right: 8px;
  }
  .icon-TipCloseFill {
    margin-right: 8px;
  }
  input::-webkit-inner-spin-button,
  input::-webkit-outer-spin-button {
    display: none;
    -webkit-appearance: none !important;
  }
}

.i-input-number-button {
  z-index: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 24px;
  height: 100%;
  border-left: 1px solid var(--i-border);
  .i-icon {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 50%;
    margin: 0;
    overflow: hidden;
    &:first-of-type {
      padding-top: 2px;
    }
    &:last-of-type {
      padding-bottom: 2px;
      border-top: 1px solid var(--i-border);
    }
    &:hover {
      background-color: var(--i-bg-disabled);
    }
    &:active {
      color: var(--i-font-4) !important;
    }
    &.i-icon-is-disabled {
      background-color: var(--i-bg-disabled);
      cursor: not-allowed;
      opacity: 0.5;
      &:hover {
        color: var(--i-font-4) !important;
      }
      &:active {
        color: var(--i-font-4) !important;
      }
    }
  }
}

.i-input-number-slider {
  position: absolute;
  left: 0;
  width: 8px;
  height: 100%;
  cursor: url('./resize.svg') 10 8, auto;
}

.i-input-number-scrubbable {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 999;
  width: auto;
  height: auto;
}

.i-input--limit {
  margin-left: 8px;
}

.i-input--limit-count {
  color: var(--i-font-5);
  font-size: 12px;
}

.i-input__inner {
  flex: 1;
  box-sizing: border-box;
  width: 100%;
  height: 32px;
  margin: 0;
  padding: 0;
  color: var(--i-font-2);
  font-size: inherit;
  line-height: 32px;
  background-color: transparent;
  border: none;
  outline: none;
  &::placeholder {
    color: var(--i-font-5);
    user-select: none;
  }
}

.i-input__group {
  display: flex;
  align-items: center;
  color: var(--i-font-2);
  .i-grid {
    flex-wrap: nowrap;
    .i-grid__item {
      .i-input {
        border-radius: 0;
        &:hover {
          box-shadow: -1px 0px 0 0px var(--i-font-4);
        }
        &:focus-within {
          box-shadow: -1px 0px 0 0px var(--i-primary);
        }
      }
      &:first-of-type {
        .i-input {
          border-radius: 4px 0 0 4px;
          &:hover,
          &:focus-within {
            box-shadow: none;
          }
        }
      }
      &:last-of-type {
        .i-input {
          border-radius: 0 4px 4px 0;
        }
      }
      &:only-of-type {
        .i-input {
          border-radius: 4px;
        }
      }
      & + .i-grid__item {
        .i-input {
          border-left-width: 0;
        }
      }
      &:focus-within {
        & + .i-grid__item {
          .i-input:hover {
            box-shadow: none;
          }
        }
      }
    }
  }
  .i-input {
    border-radius: 0;
    &:first-of-type {
      border-radius: 4px 0 0 4px;
      &:hover,
      &:focus-within {
        box-shadow: none;
      }
    }
    &:last-of-type {
      border-radius: 0 4px 4px 0;
    }
    &:only-of-type {
      border-radius: 4px;
      &:focus-within {
        box-shadow: none;
      }
    }
    & + .i-input {
      border-left-width: 0;
    }
    &:hover {
      box-shadow: -1px 0px 0 0px var(--i-font-4);
    }
    &:focus-within {
      box-shadow: -1px 0px 0 0px var(--i-primary);
      & + .i-input:hover {
        box-shadow: none;
      }
    }
  }
  .i-tag {
    height: 100%;
  }
}

.i-input__group-prefix,
.i-input__group-suffix {
  z-index: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: auto;
  height: 100%;
  padding: 0 8px;
  color: var(--i-font-4);
  font-size: 14px;
  background-color: var(--i-bg-disabled);
  border: 1px solid var(--i-border);
  user-select: none;
  &.i-input__group-cursor {
    cursor: pointer;
    transition: all 0.2s ease-out;
    &:hover {
      background-color: var(--i-bg-active);
    }
  }
}

.i-input__group-prefix {
  border-right-width: 0;
  border-radius: 4px 0 0 4px;
}

.i-input__group-suffix {
  border-left-width: 0;
  border-radius: 0 4px 4px 0;
}

.i-input__tips {
  height: auto;
  min-height: 24px;
  color: var(--i-font-5);
  font-size: 12px;
  line-height: 20px;
  &.i-input__tips--status-success {
    color: var(--i-success);
  }
  &.i-input__tips--status-warning {
    color: var(--i-warning);
  }
  &.i-input__tips--status-error {
    color: var(--i-error);
  }
}
