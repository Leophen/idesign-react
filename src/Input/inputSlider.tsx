import React from 'react';
import './index.scss';
import { InputSliderProps } from './type'

const InputSlider: React.FC<InputSliderProps> = (props) => {
  const {
    left,
    top,
    translateX,
    translateY
  } = props;

  return (
    <div
      className="i-input-number-scrubbable"
      style={{
        left: left,
        top: top,
        transform: `translate(${translateX}px,${translateY}px)`,
      }}
    >
      <svg
        width="30"
        height="19"
        viewBox="0 0 30 19"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g filter="url(#filter0_d_7775_2255)">
          <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M3 8.999V9.004L8.997 15L8.998 11.001H10.997H21V15L27 9L21 3V7.022H10.997H8.997L8.998 3L3 8.999ZM4.411 9.002L7.998 5.414L7.997 8.001H11.497H22V5.414L25.5 9L22 12.587V10L11.497 10.002L7.998 10.001L7.997 12.587L4.411 9.002Z"
            fill="white"
          />
          <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M11.4971 10.0015L22 10.0005V12.5875L25.5 9L22 5.41451V8.02151H11.4971H7.99707V5.41451L4.41107 9.0015L7.99707 12.5875V10.0005L11.4971 10.0015Z"
            fill="black"
          />
        </g>
        <defs>
          <filter
            id="filter0_d_7775_2255"
            x="-0.6"
            y="-1.6"
            width="31.2"
            height="23.2"
            filterUnits="userSpaceOnUse"
            colorInterpolationFilters="sRGB"
          >
            <feFlood floodOpacity="0" result="BackgroundImageFix" />
            <feColorMatrix
              in="SourceAlpha"
              type="matrix"
              values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"
              result="hardAlpha"
            />
            <feOffset dy="1" />
            <feGaussianBlur stdDeviation="1.3" />
            <feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.32 0" />
            <feBlend
              mode="normal"
              in2="BackgroundImageFix"
              result="effect1_dropShadow_7775_2255"
            />
            <feBlend
              mode="normal"
              in="SourceGraphic"
              in2="effect1_dropShadow_7775_2255"
              result="shape"
            />
          </filter>
        </defs>
      </svg>
    </div>
  );
};

export default InputSlider;
