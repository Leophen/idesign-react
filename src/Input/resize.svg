<svg width="22" height="19" viewBox="0 0 22 19" fill="none" xmlns="http://www.w3.org/2000/svg">
<g filter="url(#filter0_d_3664_1589)">
<path fill-rule="evenodd" clip-rule="evenodd" d="M3 8.999V9.004L8.997 15L8.998 11.001H10.997H13.017V15.001L18.997 9L13.017 3.001L13.018 7.02L10.997 7.022H8.997L8.998 3L3 8.999ZM4.411 9.002L7.998 5.414L7.997 8.001H11.497H14.018V5.416L17.583 9.002L14.019 12.587L14.018 10.002H11.497L7.998 10.001L7.997 12.587L4.411 9.002Z" fill="white"/>
<path fill-rule="evenodd" clip-rule="evenodd" d="M11.4971 10.0015H14.0181V12.5875L17.5831 9.0015L14.0181 5.41651V8.02151H11.4971H7.9971V5.41451L4.4111 9.0015L7.9971 12.5875V10.0005L11.4971 10.0015Z" fill="black"/>
</g>
<defs>
<filter id="filter0_d_3664_1589" x="-0.6" y="-1.6" width="23.2" height="23.2" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
<feFlood flood-opacity="0" result="BackgroundImageFix"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0" result="hardAlpha"/>
<feOffset dy="1"/>
<feGaussianBlur stdDeviation="1.3"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.32 0"/>
<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow_3664_1589"/>
<feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow_3664_1589" result="shape"/>
</filter>
</defs>
</svg>
