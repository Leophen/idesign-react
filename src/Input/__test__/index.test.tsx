import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from '../index';

describe('Input 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Input />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('placeholder onChange', () => {
    const changeFn = jest.fn();
    const { queryByPlaceholderText } = render(<Input placeholder='请输入内容' onChange={changeFn} />);
    expect(queryByPlaceholderText('请输入内容')).toBeInTheDocument();
    fireEvent.change(queryByPlaceholderText('请输入内容'), { target: { value: 'iDesign' } });
    expect(changeFn).toBeCalledTimes(1);
    expect(changeFn.mock.calls[0][0]).toBe('iDesign');
  });

  it('value', () => {
    const { container } = render(<Input value='iDesign' />);
    expect(container.firstChild.firstChild.value).toEqual('iDesign');
  });

  it('defaultValue', () => {
    const { container } = render(<Input defaultValue='iDesign' />);
    expect(container.firstChild.firstChild.value).toEqual('iDesign');
  });

  it('disabled', () => {
    const focusFn = jest.fn();
    const { container } = render(<Input disabled onFocus={focusFn} />);
    expect(container.firstChild).toHaveClass('i-input-is-disabled');
    fireEvent.click(container.firstChild);
    expect(focusFn).toBeCalledTimes(0);
  });

  it('readonly', () => {
    const { container } = render(<Input readonly />);
    expect(container.firstChild).toHaveClass('i-input-is-readonly');
  });

  it('size', () => {
    const { container } = render(<Input size="small" />);
    expect(container.firstChild).toHaveClass('i-input--size-small');
  })

  it('maxLength', () => {
    const { container } = render(<Input maxLength={15} />);
    expect(container.querySelector('.i-input--limit')).toBeInTheDocument();
  })

  it('password', () => {
    const { container } = render(<Input type="password" />);
    expect(container.firstChild.firstChild.type).toEqual('password');
  })

  it('number', () => {
    const { asFragment, container } = render(<Input type="number" />);
    expect(asFragment()).toMatchSnapshot();
    expect(container.firstChild.firstChild.type).toEqual('number');
  })

  it('clearable', () => {
    const clearFn = jest.fn();
    const { container } = render(<Input defaultValue='iDesign' clearable onClear={clearFn} />);
    fireEvent.mouseEnter(container.firstChild);
    expect(container.firstChild.lastChild).toHaveClass('icon-TipCloseFill');
    fireEvent.click(container.querySelector('.icon-TipCloseFill'));
    expect(container.firstChild.firstChild.value).toEqual('');
  })

  it('prefixIcon', () => {
    const { container } = render(<Input prefixIcon="Search" />);
    expect(container.firstChild.firstChild).toHaveClass('i-input-prefix-icon');
  })

  it('suffixIcon', () => {
    const { container } = render(<Input suffixIcon="Search" />);
    expect(container.firstChild.lastChild).toHaveClass('i-input-suffix-icon');
  })

  it('keyDown', async () => {
    const onKeydownFn = jest.fn();
    const onEnterFn = jest.fn();

    const { container } = render(
      <Input onKeyDown={onKeydownFn} onEnter={onEnterFn} />,
    );
    const InputDOM = container.firstChild.firstChild;

    const user = userEvent.setup()
    await user.click(InputDOM)
    await user.keyboard('abc{enter}')

    expect(onEnterFn).toBeCalled();
    expect(onKeydownFn).toBeCalled();
  });
});

