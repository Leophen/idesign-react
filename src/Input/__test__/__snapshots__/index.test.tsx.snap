// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-input"
  >
    <input
      class="i-input__inner"
      placeholder="请输入"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Input 组件测试 number 1`] = `
<DocumentFragment>
  <div
    class="i-input i-input--type-number"
  >
    <input
      class="i-input__inner"
      max="1.7976931348623157e+308"
      min="-9007199254740991"
      placeholder="请输入"
      step="1"
      type="number"
      value=""
    />
    <div
      class="i-input-number-button"
    >
      <i
        class="i-icon icon-ArrowCaretTop"
      />
      <i
        class="i-icon icon-ArrowCaretBottom"
      />
    </div>
    <div
      class="i-input-number-slider"
    />
  </div>
</DocumentFragment>
`;
